### PR TITLE
Feature/dapp build

### PR DESCRIPTION
**Pull request by hunght**
**Changes**
- added dapp:build command
- test cover for dapp:build command

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,48 +1,70 @@
...
+  });
+
+program
+  .command('dapp:build <signing-key-file>')
+  .option('-d, --dev', 'Development build mode')
+  .description('Build a DApp')
+  .action((signingKeyFile, cmd) => {
+    prompt(dappBuildSchema)
+      .then(answers => {
+        return dappBuild(answers, signingKeyFile, cmd.dev === true);
```
Comment (florianlenz) [inline index.js]: `cmd.dev === true` i think you can just pass `cmd.dev` down. The result of `cmd.dev === true` will be `true/false`.

File: src/cliActions.js [hunk 2]
```diff
@@ -6,7 +6,7 @@ const {
   decryptValue,
 } = require('./utils');
 
-const { watching } = require('./webpack');
+const { watchAndStreamingData, writeBundleFile } = require('./webpack');
```
Comment (florianlenz) [inline src/cliActions.js]: please rename `watchAndStreamingData` to `watchAndStreamData`

File: src/cliActions.js [hunk 3]
```diff
@@ -119,13 +119,12 @@ const signingKeyChangePW = (
   });
 
 /**
- * @desc Streamming Dapp
+ * @desc Check existing of signing key, and decrypt the key
```
Comment (florianlenz) [inline src/cliActions.js]: `Check existing of signing key, and decrypt the key` sorry, what do you mean? I guess it should be `check if signing key exist and decrypt it`

File: src/cliActions.js [hunk 4]
```diff
@@ -138,16 +137,47 @@ const dappStreaming = ({ pw }, signingKeyFile, devMode) =>
 
     decryptValue(signingKey.private_key_cipher_text, pw)
       // after decrypting (decrypting will fail when the password is invalid) we trigger webpack Streamming Dapp
```
Comment (florianlenz) [inline src/cliActions.js]: I don't see were we are triggering it

File: src/cliActions.test.js [hunk 5]
```diff
@@ -175,6 +176,49 @@ describe('cliActions', () => {
...
+        })
+        .catch(err => {
+          expect(err.message).toBe(
+            'Signing key ("i-do-not-exist.json") does not exist',
+          );
+          done();
+        });
+    });
+
+    test('should fail if enter wrong password', done => {
```
Comment (florianlenz) [inline src/cliActions.test.js]: should probably be `entered`

File: src/webpack.js [hunk 6]
```diff
@@ -29,6 +29,21 @@ const checkFileExistAndPromptError = path => {
     return false;
   }
 };
+
+/**
+ * @desc Make sure directory existence
```
Comment (florianlenz) [inline src/webpack.js]: `Make sure directory existence` -> `Make sure directory exists`

File: src/webpack.js [hunk 7]
```diff
@@ -29,6 +29,21 @@ const checkFileExistAndPromptError = path => {
...
+ * @param {string} filePath
+ * @return {bool}
+ */
+const ensureDirectoryExistence = filePath => {
```
Comment (florianlenz) [inline src/webpack.js]: you are not returning a bool. Since the tests are passing you didn't add a test case for this function I guess.

Comment (florianlenz) [reply src/webpack.js]: also rename the function please

File: src/webpack.js [hunk 8]
```diff
@@ -74,11 +89,63 @@ const getBuildObjectFromBundle = callback => {
 };
 
 /**
- * @desc Watching webpack build process
+ * @desc Watching and Streamming webpack build process
  * @param {bool} devMode return true when arg is --dev
  * @param {function} callback return data
  */
-const watching = (devMode, callback) => {
+const watchAndStreamingData = (devMode, callback) => {
```
Comment (florianlenz) [inline src/webpack.js]: just `watchAndStreamData`.

File: src/webpack.js [hunk 9]
```diff
@@ -74,11 +89,63 @@ const getBuildObjectFromBundle = callback => {
...
+    getBuildObjectFromBundle(callback);
+  });
+};
+
+/**
+ * @desc Write bundle file
+ * @param {bool} devMode return true when arg is --dev
+ */
+const writeBundleFile = devMode => {
+  watchingChanges(
```
Comment (florianlenz) [inline src/webpack.js]: just `watchChanges`

File: src/webpack.js [hunk 10]
```diff
@@ -74,11 +89,63 @@ const getBuildObjectFromBundle = callback => {
...
+      writeBundleBuildFile();
+    },
+    true, //stop watching
+  );
+};
+
+/**
+ * @desc Write bundle build file
+ */
+const writeBundleBuildFile = () => {
```
Comment (florianlenz) [inline src/webpack.js]: this function should only be responsible for writing data based on the name. The name is miss leading.

File: src/webpack.js [hunk 11]
```diff
@@ -74,11 +89,63 @@ const getBuildObjectFromBundle = callback => {
...
+  const dAppMetaData = getDappMetaData();
+  const filePath = path.join(process.cwd(), 'dist/index.js');
```
Comment (florianlenz) [inline src/webpack.js]: is it possible to get rid of writing all data first to `dist/index.js`? It's an not so nice solution. can't we just get the bundled code as a string from webpack and write it directly into the build file?

**Review by florianlenz — CHANGES_REQUESTED** [review]

File: src/cliActions.js [hunk 12]
```diff
@@ -119,13 +119,12 @@ const signingKeyChangePW = (
...
 /**
- * @desc Streamming Dapp
+ * @desc Check if signing key exist and decrypt it
  * @param  {string} pw password of signing key
  * @param  {string} signingKeyFile signing key file path
- * @param  {bool} devMode use dev mode to Streamming Dapp
  * @return {Promise<Promise>}
  */
-const dappStreaming = ({ pw }, signingKeyFile, devMode) =>
+const processSigningKey = ({ pw }, signingKeyFile) =>
```
Comment (florianlenz) [inline src/cliActions.js]: this function is not supposed to be in this file. This is a utility. I would also rename this function `processSigningKey` is not expressive enought

File: src/cliActions.js [hunk 13]
```diff
@@ -137,17 +136,48 @@ const dappStreaming = ({ pw }, signingKeyFile, devMode) =>
     const signingKey = JSON.parse(rawSigningKey);
 
     decryptValue(signingKey.private_key_cipher_text, pw)
-      // after decrypting (decrypting will fail when the password is invalid) we trigger webpack Streamming Dapp
+      // after decrypting (decrypting will fail when the password is invalid) return result
```
Comment (florianlenz) [inline src/cliActions.js]: I guess you can remove this comment ...

File: src/cliActions.js [hunk 14]
```diff
@@ -137,17 +136,48 @@ const dappStreaming = ({ pw }, signingKeyFile, devMode) =>
...
+      .then(res)
+      .catch(rej);
+  });
+
+/**
+ * @desc Streaming Dapp
```
Comment (florianlenz) [inline src/cliActions.js]: just `stream dapp`

File: src/cliActions.js [hunk 15]
```diff
@@ -137,17 +136,48 @@ const dappStreaming = ({ pw }, signingKeyFile, devMode) =>
...
+ * @param  {string} pw password of signing key
+ * @param  {string} signingKeyFile signing key file path
+ * @param  {bool} devMode use dev mode to Streaming Dapp
+ * @return {Promise<Promise>}
+ */
+const dappStreaming = ({ pw }, signingKeyFile, devMode) =>
```
Comment (florianlenz) [inline src/cliActions.js]: rename this as well

File: src/cliActions.js [hunk 16]
```diff
@@ -137,17 +136,48 @@ const dappStreaming = ({ pw }, signingKeyFile, devMode) =>
...
   });
 
+/**
+ * @desc Build Dapp
+ * @param  {string} pw password of signing key
+ * @param  {string} signingKeyFile signing key file path
+ * @param  {bool} devMode use dev mode to build Dapp
+ * @return {Promise<Promise>}
+ */
+const dappBuild = ({ pw }, signingKeyFile, devMode) =>
```
Comment (florianlenz) [inline src/cliActions.js]: I would call it `buildDApp`

File: src/webpack.js [hunk 17]
```diff
@@ -53,32 +67,48 @@ const getDappMetaData = () => {
 };
 
 /**
- *  @desc Get data from webpack build bundle file and combine to metadata object
- *  @param {function} callback return data with format {result:'',error:''}
+ * @desc Watching and Streaming webpack build process
```
Comment (florianlenz) [inline src/webpack.js]: first, remove the `ing`. But streaming a a build process makes no sense for me. You stream the build file ..

File: src/webpack.js [hunk 18]
```diff
@@ -53,32 +67,48 @@ const getDappMetaData = () => {
...
-    if (!err) {
-      callback({
-        result: {
-          ...dAppMetaData,
-          code: data,
-        },
-      });
-    } else {
-      callback({ error: err });
+const watchAndStreamData = (devMode, callback) => {
```
Comment (florianlenz) [inline src/webpack.js]: which data?

File: src/webpack.js [hunk 19]
```diff
@@ -53,32 +67,48 @@ const getDappMetaData = () => {
...
+  watchChanges(devMode, content => {
```
Comment (florianlenz) [inline src/webpack.js]: which changes?

File: src/webpack.js [hunk 20]
```diff
@@ -53,32 +67,48 @@ const getDappMetaData = () => {
...
+      callback({ content });
     }
   });
 };
 
+/**
+ * @desc Write bundle file
+ * @param {bool} devMode return true when arg is --dev
+ */
+const writeBundleFile = devMode => {
```
Comment (florianlenz) [inline src/webpack.js]: I think I noticed that in the last review. this function is only supposed to write a file based on the name. not to read as well.

File: src/webpack.js [hunk 21]
```diff
@@ -74,11 +89,63 @@ const getBuildObjectFromBundle = callback => {
...
+  );
+};
+
+/**
+ * @desc Write bundle build file
+ */
+const writeBundleBuildFile = () => {
+  const dAppMetaData = getDappMetaData();
+  const filePath = path.join(process.cwd(), 'dist/index.js');
+  const outputPath = path.join(process.cwd(), 'build/build.json');
```
Comment (florianlenz) [inline src/webpack.js]: yoiu don't need to join the path. You can use a relativ path `./build/build.json`

**Review by florianlenz — CHANGES_REQUESTED** [review]

File: src/webpack.js [hunk 22]
```diff
@@ -53,32 +67,48 @@ const getDappMetaData = () => {
...
+        }
+        console.log('success write update');
+      });
+    },
+    true, //stop watching
+  );
+};
+
 /**
  * @desc Watching webpack build process
```
Comment (florianlenz) [inline src/webpack.js]: just watch

File: src/cliActions.js [hunk 23]
```diff
@@ -6,7 +6,7 @@ const {
   decryptValue,
 } = require('./utils');
 
-const { watching } = require('./webpack');
+const { watchAndStreamData, writeBundleFile } = require('./webpack');
```
Comment (florianlenz) [inline src/cliActions.js]: which data?

File: src/cliActions.js (hunk 15 above)
Comment (florianlenz) [reply src/cliActions.js]: just `streamDApp`

**Review by florianlenz — CHANGES_REQUESTED** [review]

File: src/cliActions.js (hunk 12 above)
Comment (hunght) [reply src/cliActions.js]: done

File: src/cliActions.js (hunk 13 above)
Comment (hunght) [reply src/cliActions.js]: done

File: src/cliActions.js (hunk 14 above)
Comment (hunght) [reply src/cliActions.js]: done

File: src/cliActions.js (hunk 16 above)
Comment (hunght) [reply src/cliActions.js]: done

File: src/cliActions.js (hunk 15 above)
Comment (hunght) [reply src/cliActions.js]: done

File: src/webpack.js (hunk 17 above)
Comment (hunght) [reply src/webpack.js]: done

File: src/webpack.js (hunk 18 above)
Comment (hunght) [reply src/webpack.js]: fixed to watchAndStreamBundleData

File: src/webpack.js (hunk 19 above)
Comment (hunght) [reply src/webpack.js]: fixed to watchBundleChanges

File: src/webpack.js (hunk 20 above)
Comment (hunght) [reply src/webpack.js]: changed to watchAndWriteBundleFile

File: src/cliActions.js (hunk 23 above)
Comment (hunght) [reply src/cliActions.js]: change to watchAndStreamBundleData

File: src/webpack.js [hunk 24]
```diff
@@ -53,32 +67,48 @@ const getDappMetaData = () => {
...
+    devMode,
+    content => {
+      //write content to file
+      const outputPath = path.join(process.cwd(), 'build/build.json');
+      ensureDirectoryExists(outputPath);
+      fs.writeFile(outputPath, content, 'utf8', function(err) {
+        if (err) {
+          return console.log(err);
+        }
+        console.log('success write update');
```
Comment (florianlenz) [inline src/webpack.js]: this is incorrect. It should be `wrote dapp build to dapp_build.json`

File: src/webpack.js [hunk 25]
```diff
@@ -53,32 +67,48 @@ const getDappMetaData = () => {
...
+ * @param {bool} devMode return true when arg is --dev
+ */
+const writeBundleFile = devMode => {
+  watchChanges(
+    devMode,
+    content => {
+      //write content to file
+      const outputPath = path.join(process.cwd(), 'build/build.json');
+      ensureDirectoryExists(outputPath);
+      fs.writeFile(outputPath, content, 'utf8', function(err) {
```
Comment (florianlenz) [inline src/webpack.js]: can you change it to write the build to `dapp_build.json`?

**Review by florianlenz — CHANGES_REQUESTED** [review]

File: src/webpack.js (hunk 22 above)
Comment (hunght) [reply src/webpack.js]: done

File: src/webpack.js (hunk 24 above)
Comment (hunght) [reply src/webpack.js]: done

File: src/webpack.js (hunk 25 above)
Comment (hunght) [reply src/webpack.js]: done

File: src/cliActions.js [hunk 26]
```diff
@@ -119,35 +123,42 @@ const signingKeyChangePW = (
...
   });
 
+/**
+ * @desc Build Dapp
+ * @param  {string} pw password of signing key
+ * @param  {string} signingKeyFile signing key file path
+ * @param  {bool} devMode use dev mode to build Dapp
+ * @return {Promise<Promise>}
+ */
+const buildDApp = ({ pw }, signingKeyFile, devMode) =>
```
Comment (florianlenz) [inline src/cliActions.js:150]: this promise is never resolved. I am not sure if it should?

**Review by florianlenz — APPROVED** [review]

**Pull request merged**